tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/EnforceRateLimitingPlan.scala

```diff
@@ -1,70 +1,145 @@
+
+    val rateLimitedRecipients: Seq[MailAddress] = rateLimitResults.filter(_._2.equals(RateExceeded)).map(_._1)
+    val acceptableRecipients: Seq[MailAddress] = rateLimitResults.filter(_._2.equals(AcceptableRate)).map(_._1)
+
+    (acceptableRecipients, rateLimitedRecipients) match {
+      case (acceptable, _) if acceptable.isEmpty => mail.setState(exceededProcessor)
+      case (_, exceeded) if exceeded.isEmpty => // do nothing
+      case _ =>
+        mail.setRecipients(ImmutableList.copyOf(acceptableRecipients.asJava))
+
+        Using(mail.duplicate())(newMail => {
+          newMail.setRecipients(ImmutableList.copyOf(rateLimitedRecipients.asJava))
+          getMailetContext.sendMail(newMail, exceededProcessor)
+        })(LifecycleUtil.dispose(_))
+    }
+  }
+}
+
+object OperationLimitationsKey {
+  def from(rateLimitingPlanId: RateLimitingPlanId, operationLimitationsName: String): OperationLimitationsKey =
+    OperationLimitationsKey(rateLimitingPlanId.value.toString + operationLimitationsName)
 }
 
-case class PlanRateLimiterResolver @Inject()(planRepository: RateLimitationPlanRepository,
-                                             planUserRepository: RateLimitingPlanUserRepository,
-                                             rateLimiterFactory: RateLimiterFactory) {
+case class OperationLimitationsKey(value: String)
 
-  def resolve(username: Username, operationLimitationsName: String): SFlux[TmailPlanRateLimiter] =
-    SMono.fromPublisher(planUserRepository.getPlanByUser(username))
-      .flatMap(planId => SMono.fromPublisher(planRepository.get(planId)))
-      .flatMapIterable(extractRateLimiter(_, operationLimitationsName))
-
-  private def extractRateLimiter(rateLimitingPlan: RateLimitingPlan, operationLimitationsName: String): Seq[TmailPlanRateLimiter] =
-    rateLimitingPlan.operationLimitations
-      .filter(_.asString().equalsIgnoreCase(operationLimitationsName))
-      .map(aggregateRule)
-      .head
-      .map(pair => {
-        TmailPlanRateLimiter(
-          rateLimiter = rateLimiterFactory.withSpecification(pair._2, None),
-          limitTypeName = pair._1,
-          planName = rateLimitingPlan.name,
-          operationLimitations = operationLimitationsName)
-      }).toSeq
+case class PlanRateLimiterResolver(planRepository: RateLimitationPlanRepository,
+                                   rateLimiterFactory: RateLimiterFactory,
+                                   keyPrefix: Option[KeyPrefix],
+                                   precision: Option[Duration]) {
+
+  var rateLimitersMapping: Map[OperationLimitationsKey, Seq[TmailPlanRateLimiter]] = _
+
+  def init(): Unit = {
+    rateLimitersMapping = extractRateLimiters();
+  }
+
+  def retrieveRateLimiter(operationLimitationsKey: OperationLimitationsKey): Seq[TmailPlanRateLimiter] =
+    rateLimitersMapping.getOrElse(operationLimitationsKey, Seq.empty)
+
+  def extractRateLimiters(): Map[OperationLimitationsKey, Seq[TmailPlanRateLimiter]] =
```

Comment:
My idea is to use this method to:

> We can reload the rules on a periodic basis

Reply:
can we try with redis?

I would bet the limitations you encounter is linked to memoryRateLimiter1 do not have the same underlying data that memoryRateLimiter2. However with redis they shall share the same underlying data. So creating a new rate limiter 'on the fly' should be a viable option.

Reply:
> can we try with redis?

Do you worry about outdated rate-limit spec?
We can evaluate a Prefix.
'Final Prefix = Prefix In Xml + TimeStamp Slice'

> I would bet the limitations you encounter is linked to memoryRateLimiter1 do not have the same underlying data that memoryRateLimiter2.

Yes, this is the reason I use Map to store it

Reply:
> Do you worry about outdated rate-limit spec?

1. What's the impact?

Changes in plans would occur on rare occurences, if the impact is low we should not let it impact negatively our designs...

2. Worst case we can add a version number in plans and use it in the rate limiting key?

Reply:
3. We have a divergence between our two storage implementations. Either we succeed to adapt/wrap the memory one to keep the complexity within the memory rate limiting component. Or we should only rely on redis for this use case (IMO acceptable too)

In no case should we let some aspects of a specific implem (memory) negatively impact other components (here the mailet)